// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
syntax = "proto3";
import "google/protobuf/timestamp.proto";

service SampleService {
  	rpc doSimple ( Request ) returns ( Response ) {}
	rpc doResponseStreaming( Request ) returns ( stream Response ) {}
	rpc doRequestStreaming ( stream Request ) returns ( Response ) {}
	rpc doBidirectional ( stream Request ) returns ( stream Response ) {}
	rpc doSpecialDataType( SpecialDataTypeRequest ) returns ( SpecialDataTypeResponse ){}
}

message Request {
	string input = 1;
}

message Response {
	string output = 1;
}

message SpecialDataTypeRequest{
	google.protobuf.Timestamp date = 1;
	repeated string names = 2;
	map<string, string> name2phoneNumMap = 3;
	int32 numberOfCreditCard = 4;
}

message SpecialDataTypeResponse{
	google.protobuf.Timestamp date = 1;
	repeated string names = 2;
	map<string, string> name2phoneNumMap = 3;
	int32 numberOfCreditCard = 4;
}
